File: src/plugin/menus/Menu.java
package plugin.menus;

import arc.Events;
import arc.func.Cons;
import arc.struct.ObjectMap;
import arc.struct.Seq;
import mindustry.game.EventType;
import mindustry.gen.Call;
import mindustry.gen.Player;

public class Menu {
    static int lastId = 0;

    static final ObjectMap<Integer, Menu> menus = new ObjectMap<>();

    Player player;
    int id;
    Seq<Seq<String>> rows = new Seq<>();
    String title, message;
    Seq<Cons<Player>> handlers = new Seq<>();

    Menu(String title, String message) {
        this.title = title;
        this.message = message;
    }

    public Menu add(String text, Cons<Player> handler) {
        if (rows.isEmpty()) {
            rows.add(new Seq<String>());
        }

        rows.peek().add(text);

        handlers.add(handler);

        return this;
    }

    public Menu add(String text) {
        return add(text, (p) -> {});
    }

    public Menu show(Player player) {
        this.player = player;
        lastId += 1;
        this.id = lastId;

        menus.put(this.id, this);

        Call.menu(player.con, this.id, title, message, buildRows());

        return this;
    }

    public Menu row() {
        rows.add(new Seq<String>());
        return this;
    }

    String[][] buildRows() {
        String[][] rowsReturn = new String[rows.size][];

        for (int i = 0; i < rows.size; i++) {
            Seq<String> row = rows.get(i);
            rowsReturn[i] = new String[row.size];
            for (int o = 0; o < row.size; o++) {
                rowsReturn[i][o] = row.get(o);
            }
        }

        return rowsReturn;
    }

    public static void load() {
        Events.on(EventType.PlayerLeave.class, e -> {
            menus.each((id, menu) -> {
                if (menu.player == e.player) menus.remove(id);
            });
        });

        Events.on(EventType.MenuOptionChooseEvent.class, (e) -> {
            Player player = e.player;
            Menu menu = menus.get(e.menuId);
            if (menu == null || menu.player != player) {
                player.sendMessage("[scarlet]Invalid menu.");
                return;
            }

            int option = e.option;
            if (option == -1) {
                player.sendMessage("[red]Closed.");
                return;
            }
            if (option > menu.handlers.size) {
                player.sendMessage("[scarlet]Invalid option.");
                return;
            }
            menus.remove(e.menuId);
            menu.handlers.get(option).get(player);
        });
    }
}

File: src/main/kotlin/plugin/menus/Menu.kt
package plugin.menus

import arc.Events
import arc.func.Cons
import arc.struct.ObjectMap
import arc.struct.Seq
import mindustry.game.EventType
import mindustry.gen.Call
import mindustry.gen.Player
import plugin.Bundle

class Menu(var title: String, var message: String) {
    lateinit var player: Player
    var id: Int = 0
    val rows: Seq<Seq<String>> = Seq()
    val handlers: Seq<Cons<Player>> = Seq()

    fun add(text: String, handler: Cons<Player>): Menu {
        if (rows.isEmpty) {
            rows.add(Seq<String>())
        }

        rows.peek().add(text)
        handlers.add(handler)

        return this
    }

    fun add(text: String): Menu = add(text) { }

    fun show(player: Player): Menu {
        this.player = player
        lastId += 1
        this.id = lastId

        menus.put(id, this)

        Call.menu(player.con, id, resolve(title, player), resolve(message, player), buildRows(player))

        return this
    }

    fun resolve(text: String, player: Player): String {
        if (text.startsWith("@")) {
            val key = text.substring(1)
            val value = Bundle.get(key, player.locale)
            if (value != key) return value
        }

        return text
    }

    fun row(): Menu {
        rows.add(Seq<String>())
        return this
    }

    fun buildRows(player: Player): Array<Array<String>> = Array(rows.size) { i ->
        val row = rows[i]
        Array(row.size) { o -> resolve(row[o], player) }
    }

    companion object {
        var lastId = 0
        val menus = ObjectMap<Int, Menu>()

        fun load() {
            Events.on(EventType.PlayerLeave::class.java) { e ->
                val toRemove = Seq<Int>()
                menus.each { id, menu ->
                    if (menu.player == e.player) toRemove.add(id)
                }
                toRemove.each { menus.remove(it) }
            }

            Events.on(EventType.MenuOptionChooseEvent::class.java) { e ->
                val player = e.player
                val menu = menus.get(e.menuId)
                if (menu == null || menu.player != player) {
                    player.sendMessage(Bundle.get("menu.invalid", player.locale))
                    return@on
                }

                val option = e.option
                if (option == -1) {
                    player.sendMessage(Bundle.get("menu.closed", player.locale))
                    menus.remove(e.menuId)
                    return@on
                }
                if (option > menu.handlers.size) {
                    player.sendMessage(Bundle.get("menu.invalidoption", player.locale))
                    return@on
                }
                menus.remove(e.menuId)
                menu.handlers[option].get(player)
            }
        }
    }
}